rep-2017.rst

```diff
@@ -0,0 +1,359 @@
+REP: 2017
+Title: Thread attributes configuration support in rcl
+Author: Masaaki Ueno, Koichi Oyama, Shoji Morita <s-morita@esol.co.jp>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 11-Jul-2023
+Post-History: 18-Aug-2023
+
+
+Abstract
+========
+
+This REP proposed a feature of the ros client library (rcl) to accept the thread attributes, such as priorities, scheduling, and core affinity, which are to be passed to the language bindings like rclcpp.
+
+
+Motivation
+==========
+
+The configuration and setting of thread attributes for ROS 2 executor threads are essential for controlling computational resources and meeting real-time requirements, which are typically mandatory for robot systems.
+
+These attributes are particularly critical for embedded robot systems that lack redundant resources, as they aid in optimizing CPU resource utilization.
+
+However, currently, there is a lack of infrastructure to configure and set thread attributes for the threads used by the ROS 2 executors.  As a result, specific code tailored to each target environment must be implemented.  The implementation not only requires significant effort but poses some obstacles to continuous integration applied to the systems with multiple execution environments.
```

Comment:
My team will propose the usage pattern below to give thread attribution to the ROS 2 executor from the user application, as follows:

1. A user application retrieves the configured thread attribution from rcl via the newly added API proposed in this REP.
2. The user application chooses a corresponding thread attribute parameter for the executor to be launched.
3. The user application launches the executor, passing the thread attribute parameter through the executor's constructor.

My team will share the extension proposal for the existing single-thread/multi-thread executor within a few weeks in the following thread: 

https://github.com/ros2/rclcpp/pull/2205

I believe that it will be worth proposing because we currently cannot pass thread attribute parameters to ROS 2 executors.

Reply:
although this idea works to set the thread priorities to threads running inside the Executor, `that application can register the threads to ROS 2 Executors` can provide more straight-forward from user application perspective? probably we would need both to change the thread priority during runtime, i think.